### PR TITLE
Shiki perf

### LINKED ISSUE #172: Improving Shiki Language performance by Hosting Language
https://github.com/shikijs/shiki/blob/main/docs/languages.md

### PR DESCRIPTION
**Pull request by ChcJohnie**
closes #172 

utilizes fixes mentioned in issue (highlighter instance caching and explanations disabled)
**note shiki enabled in dev as with fixes perf cost seems to be negligible**

### CODE REVIEW

File: src/docs/lib/shiki.ts [hunk 1]
```diff
@@ -11,35 +11,46 @@ const theme: Theme = "github-dark";
...
+ */
 export async function getHighlighter({
   langs = defaultlangs,
 }: {
   langs?: Lang[];
-} = {}) {
-  /** Preload NO languages in development */
-  const isDevelopment = process.env.NODE_ENV === "development";
+} = {}): Promise<Highlighter> {
+  const key = [theme, ...langs].join(",");
```
Comment (iamhectorsosa) [inline src/docs/lib/shiki.ts:25]: Any reason for using a comma? Why not:

```suggestion
  const key = [theme, ...langs].join("-");
```

File: src/docs/lib/shiki.ts [hunk 2]
```diff
@@ -11,35 +11,46 @@ const theme: Theme = "github-dark";
...
+    );
 
-  /* ✅ Create a highlighter instance with a theme */
-  return await getHighlighterFromShiki({
-    theme,
-    langs: isDevelopment ? [] : langs,
-  });
+  const promise = highlighterCache.get(key);
+  if (!promise) throw new Error("Promise not found");
+  return await promise;
```
Comment (iamhectorsosa) [inline src/docs/lib/shiki.ts:38]: What's the purpose of throwing this error?

If the key isn't found in the cache, it sets a key-value. If it is found then it should just return the promise:

```suggestion
  return await highlighterCache.get(key);
```

Did you find any issues during testing?

Comment (ChcJohnie) [reply src/docs/lib/shiki.ts:38]: Actually was not sure for while, but it was to satisfy TS (get returning Promise or undefined). Wanted to check for better solution, and while there is probably no clean one, I think in this case its okay to do as we can be smart this time:
```suggestion
  return await highlighterCache.get(key)!;
```
Is it okay from your side (I was kind of guided not to overuse `!` previously but I do not see issue with it in this place)?

Comment (iamhectorsosa) [reply src/docs/lib/shiki.ts:38]: How about this?

```ts
const highlighterStore = new Map<string, Promise<Highlighter>>();

/**
 * Get a shiki highlighter instance.
 * Cached by theme and languages.
 */
export async function getHighlighter({
  langs = defaultlangs,
}: {
  langs?: Lang[];
} = {}): Promise<Highlighter> {
  const key = [theme, ...langs].join(",");
  const highlighterCache = highlighterStore.get(key);

  // If cache is found return early
  if (highlighterCache) {
    return highlighterCache;
  }

  // Create an instance of highlighter
  // Set in store and
  // Return the highlighter promise
  const highlighterPromise = getHighlighterFromShiki({
    theme,
    langs,
  });
  highlighterStore.set(
    key,
    highlighterPromise
  );
  return highlighterPromise;
}
```

Comment (ChcJohnie) [reply src/docs/lib/shiki.ts:38]: Yes, that looks quite reasonable (almost feels like I wrote this code style before 😓 ), thanks

**Review by iamhectorsosa — APPROVED** [review]

**Pull request merged**